src/input.rs

```diff
@@ -0,0 +1,191 @@
+                        self.lmb_down = false;
+                    }
+                    (winit::ElementState::Pressed, winit::MouseButton::Right) => {
+                        self.rmb_down = true;
+                    }
+                    (winit::ElementState::Released, winit::MouseButton::Right) => {
+                        self.rmb_down = false;
+                    }
+                    (_, _) => (),
+                },
+                winit::WindowEvent::Resized(logical_size) => {
+                    // Even if the window resized multiple times, only
+                    // take the last one into account.
+                    self.input_state.window_resized = Some(logical_size);
+                }
+                _ => (),
+            },
+            winit::Event::DeviceEvent { event, .. } => match event {
+                winit::DeviceEvent::MouseMotion { delta } => {
+                    if !gui_captured_mouse {
+                        let x = delta.0 as f32;
+                        let y = delta.1 as f32;
+                        if self.lmb_down && self.rmb_down {
+                            self.input_state.camera_zoom -= y;
+                        } else if self.lmb_down {
+                            self.input_state.camera_rotate[0] -= x;
+                            self.input_state.camera_rotate[1] -= y;
+                        } else if self.rmb_down {
+                            if self.shift_down {
+                                self.input_state.camera_pan_ground[0] += x;
+                                self.input_state.camera_pan_ground[1] -= y;
+                            } else {
+                                self.input_state.camera_pan_screen[0] += x;
+                                self.input_state.camera_pan_screen[1] -= y;
+                            }
+                        }
+                    }
+                }
+                winit::DeviceEvent::MouseWheel { delta, .. } => match delta {
+                    winit::MouseScrollDelta::PixelDelta(winit::dpi::LogicalPosition {
+                        y, ..
+                    }) => {
+                        if !gui_captured_mouse {
+                            match y.partial_cmp(&0.0) {
+                                Some(Ordering::Greater) => self.input_state.camera_zoom_steps += 1,
+                                Some(Ordering::Less) => self.input_state.camera_zoom_steps -= 1,
+                                _ => (),
+                            }
+                        }
+                    }
+                    winit::MouseScrollDelta::LineDelta(_, y) => {
+                        if !gui_captured_mouse {
+                            match y.partial_cmp(&0.0) {
+                                Some(Ordering::Greater) => self.input_state.camera_zoom_steps += 1,
+                                Some(Ordering::Less) => self.input_state.camera_zoom_steps -= 1,
+                                _ => (),
```

Comment:
As asked above, is this and `partial_cmp` in `winit::MouseScrollDelta::PixelDelta` prefered to simple `if / else`?